### PR TITLE
[Models] Relationships for Kubernetes ConfigMap using Environment Variables

### LINKED ISSUE #7973: [MeshModel]: Identify relationships for K8s ConfigMap
### Proposed Change

- Identify the possible relationship type for `ConfigMap`.
- Identify the related resource. 
- Define the constraints for that relationship.

---

### Contributor [Guides](https://docs.meshery.io/project/contributing) and [Handbook](https://layer5.io/community/handbook)

- [MeshModel Repository](https://github.com/meshery/meshery/tree/master/server/meshmodel)
- 🖥 [MeshModel Overview](https://docs.google.com/presentation/d/1SQMfyu5shjpGKlYONdVzOtd7UYTgLWBcgUvHMLCZ2tY/edit#slide=id.g226f5de5180_19_259)
- 🙋🏾🙋🏼 Questions: [Discussion Forum](https://discuss.layer5.io) and [Community Slack](http://slack.layer5.io)

### Instructions for Models
1. _Forthcoming_

### Instructions for Components
While the default shape for new components is a circle, each component should be considered for its best-fit shape.
1. Review and familiarize with the available set of predefined relationship types. Refer the Cytoscape [node types](https://js.cytoscape.org/demos/node-types/) for a list of possible shapes. 
1. Propose a specific shape, best-suited to visually represent the Component. Example - Deployment as a pentagon.
1. Proposee a specific icon, best-suited to visually represent the Component. Example - DaemonSet as a skull icon.

### Instructions for Relationships
1. Identify the relationship and any specific constraints to be enforced between the two specific components, their models or potentially other components, models, or environmental considerations.
1. Propose a specific visual representation for the relationship. Visual representation examples:
1.1. [Hierarchical](https://github.com/meshery/meshery/blob/master/.github/assets/images/hierarchical_relationship.png)
1.1. [Sibling](https://github.com/meshery/meshery/blob/master/.github/assets/images/sibling_relationship.png)
1.1. [Binding](https://github.com/meshery/meshery/blob/master/.github/assets/images/binding_realtionship.png)
1. Prospose the appropriate relationship type, using one of the predefined set of relationship types or suggest a new relationship where an existing type does not fit.
1. Create a Relationship Definition (yaml). See the [Relationship Schema](https://github.com/meshery/meshery/tree/master/server/meshmodel/schemas) in MeshModel and [examples]([url](https://github.com/meshery/meshery/tree/master/server/meshmodel/relationships)).
1. Create a policy for evaluation of the relationship (rego). See examples.

### Instructions for Policies
1. _Forthcoming_

### PR DESCRIPTION
**Pull request by leecalcote**
Signed-off-by: Lee Calcote <lee.calcote@layer5.io>

## Description

This PR fixes #7973 by defining a new ConfigMap relationships as an `Inventory` subtype. The following resource relationships are included:

1. Deployments
2. DaemonSet
3. ReplicaSet
4. StatefulSet
5. CronJob
6. Job 
7. Pod

Possible, but excluded relationships based on the fact that their references are implementation specific:

1. Ingress, Ingress Controllers
2. HPAs
3. Custom Controllers

### Caveats

Caveats with this approach that need review from maintainers:

1. [Easy] Hardcoded assumption of one and only one container inside each type of resources' Pod spec.
    - Less than ideal and should not be hardcoded.
    - A new UX paradigm will need to be considered by way of asking user to confirm (or providing a UI interaction) for identification of which specific (init) container(s) the ConfigMap should refer.
1. [Difficult] These relationships only consider that ConfigMaps will be made available as environment variables, not that they will be mounted as volumes. 
    - Analysis as to whether support for volume mounting ConfigMaps (example below) needs to be performed in order to assess whether volume mounted ConfigMaps can be 1) included in this same Relationship and can 2) use the same evaluation policy.

Example of volume mounting a ConfigMap (to a Pod):

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: example-pod
spec:
  containers:
  - name: init-container
    image: your/init-container-image:tag
    volumeMounts:
    - name: config-volume
      mountPath: /etc/config
  initContainers:
  - name: init-container
    image: your/init-container-image:tag
    volumeMounts:
    - name: config-volume
      mountPath: /etc/config
  - name: main-container
    image: your/main-container-image:tag
    volumeMounts:
    - name: config-volume
      mountPath: /etc/config
  volumes:
  - name: config-volume
    configMap:
      name: example-configmap
```

### CODE REVIEW

**Review by MUzairS15 — COMMENTED** [review]
1. As we build the UX to ask users what containers to apply, the policy evaluation needs to enhance and should support inputs (achievable)
2. Similar to # 1 we can ask the users whether to apply Config Map as volume or environment variable. Same relationship def can capture both of these cases, because `mutatorRef and mutatedRef` have their type as `array of arrays`. 
Based on the input received we can configure which pair of `mutatorRef` and `mutatedRef` is considered during the eval. 

IMO, an additional `operator: (and, or, onlyOne...)` attribute needs to be added, which will indicate what operation to apply. foreg: Suppose `mutatedRef` has 2 paths
`"mutatedRef": [["spec", "spec", "containers[0]", "envFrom", "configMapRef", "name"], ["spec", "spec", "containers[0]", "volumes", "configMap", "name"]],`
, for this particular case we need to choose either one but for other cases it might be required that both of these entries needs to statisfied/patched. (For eg: See existing Permission relationship).

File: server/meshmodel/kubernetes/relationships/hierarchical_inventory_configmap.json [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+          "model": "kubernetes",
+          "patch": {
+            "patchStrategy": "replace",
+            "mutatedRef": [["spec", "jobTemplate", "spec", "spec", "containers[0]", "envFrom", "configMapRef", "name"]],
+            "description": "ConfigMaps can be referenced in the Pod template specification within the CronJob definition to inject configuration data into the pods.\n\nThe keys from the ConfigMap will be exposed as environment variables to the container within the pod created by the CronJob."
+          }
+        }
+      ]
+    },
+    "deny": {
```
[the remark this answers is not in the document]
Comment (leecalcote) [reply server/meshmodel/kubernetes/relationships/hierarchical_inventory_configmap.json:98]: Deny self or deny namespace?

**Comment by leecalcote** [comment]
> 1. As we build the UX to ask users what containers to apply, the policy evaluation needs to enhance and should support inputs (achievable)
> 2. Similar to # 1 we can ask the users whether to apply Config Map as volume or environment variable. Same relationship def can capture both of these cases, because `mutatorRef and mutatedRef` have their type as `array of arrays`.
>    Based on the input received we can configure which pair of `mutatorRef` and `mutatedRef` is considered during the eval.
> 
> IMO, an additional `operator: (and, or, onlyOne...)` attribute needs to be added, which will indicate what operation to apply. foreg: Suppose `mutatedRef` has 2 paths `"mutatedRef": [["spec", "spec", "containers[0]", "envFrom", "configMapRef", "name"], ["spec", "spec", "containers[0]", "volumes", "configMap", "name"]],` , for this particular case we need to choose either one but for other cases it might be required that both of these entries needs to statisfied/patched. (For eg: See existing Permission relationship).

Are you suggesting that identification of which container is also not feasible without support for a “one of”?

**Comment by leecalcote** [comment]
@MUzairS15 commit the suggested deny.from.

**Comment by MUzairS15** [comment]
> Are you suggesting that identification of which container is also not feasible without support for a “one of”?

No, identification of container is feasible without `one of`. The idea of `operators` is for other use cases (eg: of all the conditions listed even if a majority or one of them matched then it might be a valid relationship, so to serve these type of cases.)

**Review by MUzairS15 — APPROVED** [review]
Changing the `subType` name prevents the conflict for now. Renamed  the `subType` to "Config".
Ideally the conflicts needs to be resolved by having the concept of Scopes (Global Scope: achieved via models, Local Scope: achieved via Selector set (augmenting the schema for relationship def, i.e. changing the type of `selectors` attribute to "array" from "object")

**Pull request merged**